mlir/lib/Gradient/Transforms/ConversionPatterns.cpp

```diff
@@ -117,6 +132,276 @@ struct AdjointOpPattern : public OpConversionPattern<AdjointOp> {
+
+    func::FuncOp wrappedCallee =
+        SymbolTable::lookupNearestSymbolFrom<func::FuncOp>(callee, rewriter.getStringAttr(fnName));
+    if (!wrappedCallee) {
+        PatternRewriter::InsertionGuard insertGuard(rewriter);
+        rewriter.setInsertionPointAfter(callee);
+
+        wrappedCallee = rewriter.create<func::FuncOp>(loc, fnName, fnType, visibility);
+        Block *entryBlock = wrappedCallee.addEntryBlock();
+        rewriter.setInsertionPointToStart(entryBlock);
+
+        // Get the arguments
+        SmallVector<Value> callArgs(wrappedCallee.getArguments().begin(),
+                                    wrappedCallee.getArguments().end() - callee.getNumResults());
+
+        for (auto [arg, originalType] : llvm::zip(callArgs, originalArgTypes)) {
+            if (arg.getType().dyn_cast<LLVM::LLVMPointerType>()) {
+                Value memrefStruct = rewriter.create<LLVM::LoadOp>(loc, arg);
+                arg = rewriter.create<UnrealizedConversionCastOp>(loc, originalType, memrefStruct)
+                          .getResult(0);
+            }
+        }
+        // Call the callee
+        ValueRange results = rewriter.create<func::CallOp>(loc, callee, callArgs).getResults();
+        
+        results = rewriter.create<UnrealizedConversionCastOp>(loc, convertedResTypes, results)
+                      .getResults();
+        ValueRange resArgs = wrappedCallee.getArguments().drop_front(callee.getNumArguments());
+
+        // Store the results
+        for (auto [result, resArg] : llvm::zip(results, resArgs)) {
+            rewriter.create<LLVM::StoreOp>(loc, result, resArg);
+        }
+
+        // Add return op
+        rewriter.create<func::ReturnOp>(loc);
+    }
+
+    return wrappedCallee;
+}
+
+struct BackpropOpPattern : public OpConversionPattern<BackpropOp> {
+    using OpConversionPattern::OpConversionPattern;
+
+    LogicalResult matchAndRewrite(BackpropOp op, BackpropOpAdaptor adaptor,
+                                  ConversionPatternRewriter &rewriter) const override
+    {
+        Location loc = op.getLoc();
+        MLIRContext *ctx = getContext();
+        LLVMTypeConverter llvmTypeConverter(ctx);
+
+        Type vectorType =
+            llvmTypeConverter.convertType(MemRefType::get({UNKNOWN}, Float64Type::get(ctx)));
+
+        for (Type type : op.getResultTypes()) {
+            if (!type.isa<MemRefType>())
+                return op.emitOpError("must be bufferized before lowering");
+        }
+
+        // The callee of the backprop Op
+        func::FuncOp callee =
+            SymbolTable::lookupNearestSymbolFrom<func::FuncOp>(op, op.getCalleeAttr());
+        assert(callee);
+
+        // Creat constants
+        Value c0 = rewriter.create<LLVM::ConstantOp>(loc, rewriter.getI32IntegerAttr(0));
+        Value c1 = rewriter.create<LLVM::ConstantOp>(loc, rewriter.getI32IntegerAttr(1));
+
+        // Create mlir memref to llvm
+        StringRef allocFnName = "_mlir_memref_to_llvm_alloc";
```

Comment:
What does this function do?

Reply:
This function is provided by the flag: 

```
        --use-generic-functions                        - Use generic allocation and deallocation functions instead of the classic 'malloc', 'aligned_alloc' and 'free' functions
```

which is found in `mlir-opt` and others.

The use of this function is to replace the call to malloc when lowering `memref::AllocOp`s to allow for flexibility. We had some issues with deallocation in the past. We essentially manage the memory in the runtime in a manner similar to a memory arena. All allocations in the JIT compiled function are captured and will be freed once the JIT compiled function returns.